source/extensions/access_loggers/grpc/http_grpc_access_log_impl.h

```diff
@@ -51,12 +51,12 @@ class HttpGrpcAccessLog : public Common::ImplBase {
 
   Stats::Scope& scope_;
   const envoy::extensions::access_loggers::grpc::v3::HttpGrpcAccessLogConfig config_;
-  const ThreadLocal::SlotPtr tls_slot_;
   const GrpcCommon::GrpcAccessLoggerCacheSharedPtr access_logger_cache_;
   std::vector<Http::LowerCaseString> request_headers_to_log_;
   std::vector<Http::LowerCaseString> response_headers_to_log_;
   std::vector<Http::LowerCaseString> response_trailers_to_log_;
   std::vector<std::string> filter_states_to_log_;
+  const ThreadLocal::SlotPtr tls_slot_; // Destruct first to flush callbacks.
```

Comment:
I think this type of change can likely be removed via also deleting the still alive guard so it can't be locked. Then the callbacks would flush out but wouldn't actually be executed. Although, now that I'm typing this, I'm wondering if all of this can be simplified to just have the callbacks not run at all once the slot goes away. I will think about this more tomorrow.

Reply:
Oh nevermind, I know why I didn't do this, because the shared_ptr deletion can race within the lambda, and there is still a problem.